### ISSUE TITLE
AWS: node lifecycle controller occasionally removes k8s nodes from the cluster immediately after startup

### ISSUE THREAD

**Issue by mwhittington21** [kubernetes/kubernetes#81636]
**What happened**:
Occasionally, when a new node in one of our busy clusters starts up it will be deleted by `kube-controller-manager` within the first minute of the Kubelet on that node starting and registering with the k8s API.

In the following logs, the node name has been replaced with `<hostname>` for brevity.

The `kube-controller-manager` logs the following:
```
7:12:35.844 AM: Event(v1.ObjectReference{Kind:"Node", Namespace:"", Name:"<hostname>", UID:"bbf32ed3-c133-11e9-b67d-0a6c12c9e9b2", APIVersion:"", ResourceVersion:"", FieldPath:""}): type: 'Normal' reason: 'RegisteredNode' Node <hostname> event: Registered Node <hostname> in Controller
7:12:36.275 AM: Event(v1.ObjectReference{Kind:"Node", Namespace:"", Name:"<hostname>", UID:"bbf32ed3-c133-11e9-b67d-0a6c12c9e9b2", APIVersion:"", ResourceVersion:"", FieldPath:""}): type: 'Normal' reason: 'DeletingNode' Node <hostname> event: Deleting Node <hostname> because it's not present according to cloud provider
```

Note that there is .4 of a second between the node being registered and then deleted.

The `kubelet` on the node logs the following after being evicted, and until it is manually restarted:
```
<hostname> kubelet_v1.13.7[1894]: E0820 04:19:32.848133    1894 kubelet_node_status.go:381] Error updating node status, will retry: error getting node "<hostname>": nodes "<hostname>" not found
<hostname> kubelet_v1.13.7[1894]: E0820 04:19:32.848332    1894 kubelet_node_status.go:368] Unable to update node status: update node status exceeds retry count
```

**What you expected to happen**:
The node should always join the cluster and never be automatically evicted by `kube-controller-manager` while it still exists.

**How to reproduce it (as minimally and precisely as possible)**:
We haven't been able to reproduce it, but we observe 4-8 cases of this per week with a total node churn in the thousands.

**Anything else we need to know?**:
It is our current suspicion that this bug occurs when the AWS EC2 API isn't completely up to date at the time the Kubelet starts. As an eventually consistent API, the box is actually already started and running via it's AutoScaling Group, but the `kube-controller-manager` cannot find it.

The `kube-controller-manager` will then prune the node from the API because it no longer exists. The code where this pruning occurs is [here](https://github.com/kubernetes/kubernetes/blob/v1.13.10/pkg/controller/cloud/node_controller.go#L295) (or [here](https://github.com/kubernetes/kubernetes/blob/v1.15.3/pkg/controller/cloud/node_lifecycle_controller.go#L180) in v1.14 onwards).

Our node startup time is less than 1 minute from cold boot to Kubelet start.

We have a high node churn in our clusters as we run primarily job-based workloads. We scale up the number of nodes to cope with the jobs, and scale down when we have processed most of the load. This scaling is automated. Some clusters hit 500+ nodes at maximum load. This node churn may contribute to this problem we are having.

**Environment**:
- Kubernetes version (use `kubectl version`): 1.13.7 & 1.13.9
- Cloud provider or hardware configuration: AWS
- OS (e.g: `cat /etc/os-release`): Container Linux by CoreOS 2135.4.0 (Rhyolite)
- Kernel (e.g. `uname -a`): Linux <hostname> 4.19.50-coreos-r1 #1 SMP Mon Jun 24 22:08:34 -00 2019 x86_64 Intel(R) Xeon(R) Platinum 8175M CPU @ 2.50GHz GenuineIntel GNU/Linux
- Others:

**Comment by mwhittington21** [comment]
@kubernetes/sig-cloud-provider-aws

**Comment by mwhittington21** [comment]
@kubernetes/sig-cloud-provider-bugs

/sig cloud-provider

**Comment by M00nF1sh** [comment]
@mwhittington21 
Is this issue still happening? What's your VPC DNS setting and did you have custom DHCP settings.
BTW, would you help check whether the spec.ProviderID is set on your nodes?

**Comment by qqshfox** [comment]
We are facing this exact issue in our production cluster using 1.11.10.

Here're the KCM logs:

```
W1219 08:59:07.552023       1 actual_state_of_world.go:491] Failed to update statusUpdateNeeded field in actual state of world: Failed to set statusUpdateNeeded to needed true, because nodeName="ip-172-31-43-4.cn-north-1.compute.internal" does not exist
I1219 08:59:07.559897       1 range_allocator.go:310] Set node ip-172-31-43-4.cn-north-1.compute.internal PodCIDR to 100.115.202.0/24
I1219 08:59:07.570534       1 ttl_controller.go:271] Changed ttl annotation for node ip-172-31-43-4.cn-north-1.compute.internal to 0 seconds
I1219 08:59:07.572674       1 ttl_controller.go:271] Changed ttl annotation for node ip-172-31-43-4.cn-north-1.compute.internal to 0 seconds
I1219 08:59:13.184448       1 node_lifecycle_controller.go:558] Controller observed a new Node: "ip-172-31-43-4.cn-north-1.compute.internal"
I1219 08:59:13.184455       1 controller_utils.go:212] Recording Registered Node ip-172-31-43-4.cn-north-1.compute.internal in Controller event message for node ip-172-31-43-4.cn-north-1.compute.internal
I1219 08:59:13.870874       1 node_lifecycle_controller.go:712] Deleting node (no longer present in cloud provider): ip-172-31-43-4.cn-north-1.compute.internal
I1219 08:59:13.870890       1 controller_utils.go:212] Recording Deleting Node ip-172-31-43-4.cn-north-1.compute.internal because it's not present according to cloud provider event message for node ip-172-31-43-4.cn-north-1.compute.internal
I1219 08:59:19.149720       1 node_lifecycle_controller.go:570] Controller observed a Node deletion: ip-172-31-43-4.cn-north-1.compute.internal
I1219 08:59:19.149723       1 controller_utils.go:212] Recording Removing Node ip-172-31-43-4.cn-north-1.compute.internal from Controller event message for node ip-172-31-43-4.cn-north-1.compute.internal
I1219 08:59:25.253194       1 gc_controller.go:162] Found orphaned Pod prom-prometheus-node-exporter-rmjhg assigned to the Node ip-172-31-43-4.cn-north-1.compute.internal. Deleting.
I1219 08:59:25.278278       1 gc_controller.go:162] Found orphaned Pod k8s-spot-termination-handler-rg5bh assigned to the Node ip-172-31-43-4.cn-north-1.compute.internal. Deleting.
I1219 08:59:25.331078       1 gc_controller.go:162] Found orphaned Pod calico-node-b6fsp assigned to the Node ip-172-31-43-4.cn-north-1.compute.internal. Deleting.
I1219 08:59:25.445601       1 gc_controller.go:162] Found orphaned Pod kube-proxy-ip-172-31-43-4.cn-north-1.compute.internal assigned to the Node ip-172-31-43-4.cn-north-1.compute.internal. Deleting.
I1219 08:59:25.445612       1 gc_controller.go:62] PodGC is force deleting Pod: kube-system:kube-proxy-ip-172-31-43-4.cn-north-1.compute.internal
I1219 08:59:25.471307       1 gc_controller.go:166] Forced deletion of orphaned Pod kube-proxy-ip-172-31-43-4.cn-north-1.compute.internal succeeded
I1219 08:59:25.572101       1 gc_controller.go:162] Found orphaned Pod prometheus-operator-prometheus-node-exporter-8cpcx assigned to the Node ip-172-31-43-4.cn-north-1.compute.internal. Deleting.
I1219 08:59:25.593833       1 gc_controller.go:162] Found orphaned Pod node-problem-detector-vn6jc assigned to the Node ip-172-31-43-4.cn-north-1.compute.internal. Deleting.
I1219 08:59:25.748033       1 gc_controller.go:162] Found orphaned Pod fluentd-kinesis-lrr7h assigned to the Node ip-172-31-43-4.cn-north-1.compute.internal. Deleting.
```

New node registered at 08:59:13, deleted 6s later at 08:59:19.

From these logs, the actual delete logic, which is at https://github.com/kubernetes/kubernetes/blob/v1.11.10/pkg/controller/nodelifecycle/node_lifecycle_controller.go#L706, is not using `ProviderID` at all.

> By default, kubelet uses describeInstance with private-dns setting to nodeName to figure out instanceID and set as spec.ProviderId. there can be some bugs around it when there are custom vpc settings.

I believe the logic @M00nF1sh mentioned is https://github.com/kubernetes/kubernetes/blob/v1.11.10/pkg/controller/cloud/node_controller.go#L274.

**Comment by qqshfox** [comment]
Deployed a modified KCM into our production cluster last night, we can't reproduce this issue any more by now.

There were at least 5 nodes couldn't got deleted before the patch when spawning 50 nodes. Now we tried spawning 100 nodes for more than 10 times, all nodes successfully registered.

We'll let the modified KCM keep running for some days, to see if this problem will appear again.

**Comment by qqshfox** [comment]
xref #58334

**Comment by qqshfox** [comment]
> Deployed a modified KCM into our production cluster last night, we can't reproduce this issue any more by now.
> 
> There were at least 5 nodes got deleted before the patch when spawning 50 nodes. Now we tried spawning 100 nodes for more than 10 times, all nodes successfully registered.
> 
> We'll let the modified KCM keep running for some days, to see if this problem will appear again.

We have not yet experienced this issue since we deployed the modified KCM into our production cluster 3 weeks ago. We upgraded our cluster from 1.11 to 1.13. For each minor version we upgraded to, we deployed a patched KCM and running for a whole week.

However, we haven't tested and haven't inspected the later version, e.g. 1.14, which rewrote this part from https://github.com/kubernetes/kubernetes/blob/v1.11.10/pkg/controller/nodelifecycle/node_lifecycle_controller.go#L706 to https://github.com/kubernetes/kubernetes/blob/v1.14.10/pkg/controller/nodelifecycle/node_lifecycle_controller.go#L802. Since the next following weeks are the Chinese New Year, we'll keep the 1.13 cluster running for a longer time. Hopefully we can get some time to test the 1.14 version when we get back to work.

**Comment by qqshfox** [comment]
> > Deployed a modified KCM into our production cluster last night, we can't reproduce this issue any more by now.
> > There were at least 5 nodes got deleted before the patch when spawning 50 nodes. Now we tried spawning 100 nodes for more than 10 times, all nodes successfully registered.
> > We'll let the modified KCM keep running for some days, to see if this problem will appear again.
> 
> We have not yet experienced this issue since we deployed the modified KCM into our production cluster 3 weeks ago. We upgraded our cluster from 1.11 to 1.13. For each minor version we upgraded to, we deployed a patched KCM and running for a whole week.
> 
> However, we haven't tested and haven't inspected the later version, e.g. 1.14, which rewrote this part from https://github.com/kubernetes/kubernetes/blob/v1.11.10/pkg/controller/nodelifecycle/node_lifecycle_controller.go#L706 to https://github.com/kubernetes/kubernetes/blob/v1.14.10/pkg/controller/nodelifecycle/node_lifecycle_controller.go#L802. Since the next following weeks are the Chinese New Year, we'll keep the 1.13 cluster running for a longer time. Hopefully we can get some time to test the 1.14 version when we get back to work.

We didn't have this issue since we were running a patched 1.13 cluster in production env for more than 2 months.
Two weeks ago, we upgraded the cluster to vanilla 1.14 and have been running for two whole weeks. We still have not encountered this issue yet.

Just want to share the results with you guys here.

**Comment by diversario** [comment]
@qqshfox would you be able to share the KCM patch used in your clusters?

**Comment by qqshfox** [comment]
> @qqshfox would you be able to share the KCM patch used in your clusters?

The patches are here:

* 1.11: https://github.com/qqshfox/kubernetes/tree/release-1.11-alo7
* 1.12: https://github.com/qqshfox/kubernetes/tree/release-1.12-alo7
* 1.13: https://github.com/qqshfox/kubernetes/tree/release-1.13-alo7

**Comment by andrewsykim** [comment]
Can we close this issue? Sounds like the fix is already in v1.14 but it wasn't backported cause those versions are unsupported.
